IValue stores undefined tensor as None

smessmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #42043 Remove argument unpacking from variable kernels
* **#42025 IValue stores undefined tensor as None**

This makes sure that IValue stores None and undefined tensor in the same way,
eliminating the weird case where you do IValue.to<optional<Tensor>>() and you
don't get a None but a Some(undefined tensor).

Differential Revision: [D22733006](https://our.internmc.facebook.com/intern/diff/D22733006/)

smessmer committed:
IValue stores undefined tensor as None

This makes sure that IValue stores None and undefined tensor in the same way,
eliminating the weird case where you do IValue.to<optional<Tensor>>() and you
don't get a None but a Some(undefined tensor).

Differential Revision: [D22733006](https://our.internmc.facebook.com/intern/diff/D22733006/)

[ghstack-poisoned]

smessmer committed:
Update on "IValue stores undefined tensor as None"


This makes sure that IValue stores None and undefined tensor in the same way,
eliminating the weird case where you do IValue.to<optional<Tensor>>() and you
don't get a None but a Some(undefined tensor).

Differential Revision: [D22733006](https://our.internmc.facebook.com/intern/diff/D22733006/)

[ghstack-poisoned]

smessmer committed:
Update on "IValue stores undefined tensor as None"


This makes sure that IValue stores None and undefined tensor in the same way,
eliminating the weird case where you do IValue.to<optional<Tensor>>() and you
don't get a None but a Some(undefined tensor).

Differential Revision: [D22733006](https://our.internmc.facebook.com/intern/diff/D22733006/)

[ghstack-poisoned]

smessmer committed:
Update on "IValue stores undefined tensor as None"


This makes sure that IValue stores None and undefined tensor in the same way,
eliminating the weird case where you do IValue.to<optional<Tensor>>() and you
don't get a None but a Some(undefined tensor).

Differential Revision: [D22733006](https://our.internmc.facebook.com/intern/diff/D22733006/)

[ghstack-poisoned]

ezyang commented on aten/src/ATen/core/ivalue.h:
```diff
@@ -248,14 +248,22 @@ struct CAFFE2_API IValue final {
+      // is tagged as a tensor, is_intrusive_ptr is set to false.
+      // This is not an optional optimization: our incref call
+      // *will not* do the right thing when called on an
+      // undefined tensor.
```
This comment is stale now

ezyang commented:
This seems reasonable enough but we should loop in someone from JIT side on this PR.

smessmer committed:
Update on "IValue stores undefined tensor as None"


This makes sure that IValue stores None and undefined tensor in the same way,
eliminating the weird case where you do IValue.to<optional<Tensor>>() and you
don't get a None but a Some(undefined tensor).

Differential Revision: [D22733006](https://our.internmc.facebook.com/intern/diff/D22733006/)

[ghstack-poisoned]

smessmer committed:
Update on "IValue stores undefined tensor as None"


This makes sure that IValue stores None and undefined tensor in the same way,
eliminating the weird case where you do IValue.to<optional<Tensor>>() and you
don't get a None but a Some(undefined tensor).

Differential Revision: [D22733006](https://our.internmc.facebook.com/intern/diff/D22733006/)

[ghstack-poisoned]

smessmer committed:
Update on "IValue stores undefined tensor as None"


This makes sure that IValue stores None and undefined tensor in the same way,
eliminating the weird case where you do IValue.to<optional<Tensor>>() and you
don't get a None but a Some(undefined tensor).

Differential Revision: [D22733006](https://our.internmc.facebook.com/intern/diff/D22733006/)

[ghstack-poisoned]

suo commented on aten/src/ATen/core/ivalue.h:
```diff
@@ -248,14 +248,22 @@ struct CAFFE2_API IValue final {
+      // undefined tensor.
+      payload.as_intrusive_ptr = t.unsafeReleaseTensorImpl();
+    } else {
+      // If the tensor was undefined, set the IValue to None instead
```
This means that undefined tensors no longer return true for `isTensor()`, which seems a little confusing:
```
auto iv = IValue(at::Tensor())
iv.isTensor()  // false???
```
This may also have wide-reaching implications. For example, in `constant_propagation.cpp`, I see code that looks like:
```
    if (v.isTensor()) {
      at::Tensor t = v.toTensor();
      if (t.defined() && t.requires_grad()) {
        ...
```
with this change, the initial check for `isTensor` will fail, causing the whole block to to be skipped. I'm not sure if it matters, but we'll have to check all the callsites for `isTensor` to make sure they do not care about this change.

smessmer replied:
ugh. How much do you think we can rely on CI here?

smessmer committed:
Update on "IValue stores undefined tensor as None"


This makes sure that IValue stores None and undefined tensor in the same way,
eliminating the weird case where you do IValue.to<optional<Tensor>>() and you
don't get a None but a Some(undefined tensor).

Differential Revision: [D22733006](https://our.internmc.facebook.com/intern/diff/D22733006/)

[ghstack-poisoned]

smessmer committed:
Update on "IValue stores undefined tensor as None"


This makes sure that IValue stores None and undefined tensor in the same way,
eliminating the weird case where you do IValue.to<optional<Tensor>>() and you
don't get a None but a Some(undefined tensor).

Differential Revision: [D22733006](https://our.internmc.facebook.com/intern/diff/D22733006/)

[ghstack-poisoned]

smessmer committed:
Update on "IValue stores undefined tensor as None"


This makes sure that IValue stores None and undefined tensor in the same way,
eliminating the weird case where you do IValue.to<optional<Tensor>>() and you
don't get a None but a Some(undefined tensor).

Differential Revision: [D22733006](https://our.internmc.facebook.com/intern/diff/D22733006/)

[ghstack-poisoned]

smessmer committed:
Update on "IValue stores undefined tensor as None"


This makes sure that IValue stores None and undefined tensor in the same way,
eliminating the weird case where you do IValue.to<optional<Tensor>>() and you
don't get a None but a Some(undefined tensor).

Differential Revision: [D22733006](https://our.internmc.facebook.com/intern/diff/D22733006/)

[ghstack-poisoned]

smessmer committed:
Update on "IValue stores undefined tensor as None"


This makes sure that IValue stores None and undefined tensor in the same way,
eliminating the weird case where you do IValue.to<optional<Tensor>>() and you
don't get a None but a Some(undefined tensor).

Differential Revision: [D22733006](https://our.internmc.facebook.com/intern/diff/D22733006/)

[ghstack-poisoned]

smessmer committed:
Update on "IValue stores undefined tensor as None"


This makes sure that IValue stores None and undefined tensor in the same way,
eliminating the weird case where you do IValue.to<optional<Tensor>>() and you
don't get a None but a Some(undefined tensor).

Differential Revision: [D22733006](https://our.internmc.facebook.com/intern/diff/D22733006/)

[ghstack-poisoned]
